sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ProtonSession.java

```diff
@@ -0,0 +1,451 @@
+                    .log("session local open scheduled.");
+
+                if (state.compareAndSet(State.EMPTY, new State(session))) {
+                    openAwaiter.emitEmpty(FAIL_FAST);
+                } else {
+                    session.close();
+                    if (state.get() == State.DISPOSED) {
+                        openAwaiter.emitError(new ProtonSessionClosedException(NOT_OPENING_DISPOSED_SESSION),
+                            FAIL_FAST);
+                    } else {
+                        openAwaiter.emitError(new IllegalStateException("session is already opened."), FAIL_FAST);
+                    }
+                }
+            });
+        } catch (Exception e) {
+            if (e instanceof IOException | e instanceof RejectedExecutionException) {
+                final String message = String.format(REACTOR_CLOSED_MESSAGE_FORMAT, getConnectionId(), getName());
+                openAwaiter.emitError(retriableAmqpError(null, message, e), FAIL_FAST);
+            } else {
+                openAwaiter.emitError(e, FAIL_FAST);
+            }
+        }
+        return openAwaiter.asMono();
+    }
+
+    /**
+     * Gets a bidirectional channel on the session.
+     * <p>
+     * A channel consists of a sender link and a receiver link, and is used for management operations where a sender
+     * link is used to send a request to the broker and a receiver link gets the associated response from the broker.
+     * </p>
+     *
+     * @param name the channel name, which is used as the name prefix for sender link and receiver link in the channel.
+     * @param timeout the timeout for obtaining the channel.
+     *
+     * @return a mono that completes with a {@link ProtonChannel} once the channel is created in the session.
+     * <p>
+     *  <ul>
+     *      <li>the mono terminates with {@link IllegalStateException} if the session is not opened yet via {@link #open()}.</li>
+     *      <li>the mono terminates with
+     *          <ul>
+     *              <li>a retriable {@link ProtonSessionClosedException} if the session is disposed,</li>
+     *              <li>a retriable {@link AmqpException} if obtaining channel timeout,</li>
+     *              <li>or a retriable {@link AmqpException} if the connection reactor thread is shutdown.</li>
+     *          </ul>
+     *      </li>
+     *  </ul>
+     * </p>
+     */
+    Mono<ProtonChannel> channel(final String name, Duration timeout) {
+        final Mono<ProtonChannel> channel = Mono.create(sink -> {
+            if (name == null) {
+                sink.error(new NullPointerException("'name' cannot be null."));
+                return;
+            }
+            try {
+                // Pre-check to fail fast before using reactor dispatcher thread. The pre-check asserts that the session
+                // is opened and not in disposed state.
+                getSession("channel");
```

Comment:
nit: it'd be more readable  if we had a method like

```java
    private Throwable checkState(State s) {
        Throwable error = null;
        if (s == State.EMPTY) {
            error  = logger.logExceptionAsError(new IllegalStateException(SESSION_NOT_OPENED));
        }
        if (s == State.DISPOSED) {
            error = logger.logExceptionAsWarning(
                new ProtonSessionClosedException(String.format(DISPOSED_MESSAGE_FORMAT, resourceType)));
        }
        return error;
    }
```

then did 

```java
Throwable error = checkState(state.get();`
if (error != null) {
  sink(error);
  return;
}
```

It'd make it clear that there are no side-effects in this operation. Also we'd not use exception handling for logic.

Also could it still be possible for an error to happen before we do `getSession` on the line 238? If so, why do we need this one?

Reply:
Thank you, yes, a separate method like the one you proposed will improve the readability.

Reply:
The reason for doing this check on entry is to see if we can "fail fast in the current thread", before scheduling the work to Qpid-thread. So yes, we can still encounter error by the time Qpid-thread picks the work, but we can save thread-hopping-cost with that entry check.

Reply:
reworked this type based on the feedback.